pkg/serviceaccount/jwt.go

```diff
@@ -225,22 +225,59 @@ func (j *jwtTokenGenerator) GenerateToken(claims *jwt.Claims, privateClaims inte
 // JWTTokenAuthenticator authenticates tokens as JWT tokens produced by JWTTokenGenerator
 // Token signatures are verified using each of the given public keys until one works (allowing key rotation)
 // If lookup is true, the service account and secret referenced as claims inside the token are retrieved and verified with the provided ServiceAccountTokenGetter
-func JWTTokenAuthenticator(issuers []string, keys []interface{}, implicitAuds authenticator.Audiences, validator Validator) authenticator.Token {
+func JWTTokenAuthenticator(issuers []string, keyGetter PublicKeyGetter, implicitAuds authenticator.Audiences, validator Validator) authenticator.Token {
 	issuersMap := make(map[string]bool)
 	for _, issuer := range issuers {
 		issuersMap[issuer] = true
 	}
 	return &jwtTokenAuthenticator{
 		issuers:      issuersMap,
-		keys:         keys,
+		keys:         keyGetter,
 		implicitAuds: implicitAuds,
 		validator:    validator,
 	}
 }
 
+// PublicKeyGetter returns public keys for a given key id.
+type PublicKeyGetter interface {
+	// GetPublicKeys returns public keys to use for verifying a token with the given key id.
+	// The token keyID may be empty if the token did not have a key id header.
+	GetPublicKeys(keyID string) []interface{}
+}
+
+type staticPublicKeyGetter struct {
+	publicKeys     []interface{}
+	publicKeysByID map[string][]interface{}
+}
+
+// StaticPublicKeyGetter constructs an implementation of PublicKeyGetter
+// which returns all public keys when key id is unspecified, and returns
+// the public keys matching the keyIDFromPublicKey-derived key id when
+// a key id is specified.
+func StaticPublicKeyGetter(publicKeys []interface{}) *staticPublicKeyGetter {
+	publicKeysByID := map[string][]interface{}{}
+	for _, key := range publicKeys {
+		keyID, err := keyIDFromPublicKey(key)
+		if err == nil {
```

Comment:
I feel like this should `panic` when `err != nil` because at both call sites where this is used, the input is always valid.

Reply:
changed the signature to return an error and handled that at call sites